### PR TITLE
Add the general architecture description of the neural token parser

### PR DESCRIPTION
**Pull request by warenlg**

### CODE REVIEW

File: sourced/ml/algorithms/neural_splitter_arch.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# Neural Token Splitter architecture description
+
+The goal of this feature is to train a neural network to learn how to split identifiers, the rule-based `TokenParser` is not able to process such as "`foobar`" or "`methodbase`". Except specifc classes needed, the code is going to be stored in the algorithms directory: Class: `NeuralTokenSplitter`. The architecture will follow the following steps.
+
+1. Use `utils.engine.create_engine` to initialize a spark session and the engine with `args.repositories` as input, siva format. The Dataset is located in our cluster `/apps/borges/root-repositories`: >150k top starred repos, any languages.
+
+2. Code a new class `CodeExtractor` in `transformers.basic` that is going to return both the source codes as strings and the languages.
+
+3. Get the Token stream of identifiers thanks to [pygments.highlight](http://pygments.org/docs/quickstart/) with the `RawFormatter` and the language output by [enry](https://github.com/src-d/enry) as a lexer. Filter the token types with a `callback=filter_tokens`
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/neural_splitter_arch.md]: Design the code to easily switch between Babelfish and pygments

File: sourced/ml/algorithms/neural_splitter_arch.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+
+4. Build the training dataset
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/neural_splitter_arch.md]: Collect ~build~

File: sourced/ml/algorithms/neural_splitter_arch.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+    * Select all identifiers which return non 1-length lists when splitting them according to special charaters such as _, -, uppercase letters and more generally `NAME_BREAKUP_RE = re.compile(r"[^a-zA-Z]+")`. Ex `foo_bar` and `methodBase`
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/neural_splitter_arch.md]: if splitted on special characters (`[^a-zA-Z]+`) or case changes

File: sourced/ml/algorithms/neural_splitter_arch.md [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+    * Use `TokenParser` from `algorithms.token_parser` to split them, make them lowercase, and join them again. Ex `foobar` and `methodbase`
+
+    * We have X and y:
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/neural_splitter_arch.md]: X and Y

or

x and y

File: sourced/ml/algorithms/neural_splitter_arch.md [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+        * `foobar` -> `foo bar`
+        * `methodbase` -> `method base`
+
+5. Train a basic Tensorflow neural net directly in the cluster, to learn how to split those identifiers
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/neural_splitter_arch.md]: Which arch? Where is the reference code?

File: sourced/ml/algorithms/neural_splitter_arch.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+6. Evaluation plots and accuracy improvments by playing with the global architecture. Ideas:
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/neural_splitter_arch.md]: playing with metaparameters. Other ideas:

File: sourced/ml/algorithms/neural_splitter_arch.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+    * Use the context of identifiers
+
+    * Focus on Python, Java repositories and use [Babelfish](https://github.com/bblfsh)
```
Comment (vmarkovtsev) [inline sourced/ml/algorithms/neural_splitter_arch.md]: Python also uses Babelfish

File: doc/neural_splitter_arch.md [hunk 8]
```diff
@@ -0,0 +1,29 @@
+# Neural Token Splitter architecture description
+
+The goal of this feature is to train a neural network to learn how to split identifiers. Indeed, the rule-based `TokenParser` is not able to process some of them such as "`foobar`" or "`methodbase`". The main code is going to be stored in the algorithms directory: Class: `NeuralTokenSplitter`. The implementation plan is a follows.
```
Comment (vmarkovtsev) [inline doc/neural_splitter_arch.md]: is **as** follows

File: doc/neural_splitter_arch.md [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+
+    * Use `TokenParser` from `algorithms.token_parser` to split them, make them lowercase, and join them again. Ex `foobar` and `methodbase`
+
+    * We have X and Y:
+        * `foobar` -> `foo bar`
+        * `methodbase` -> `method base`
+
+5. Code a simple neural language model with Keras that relies on character-level inputs as it is described in [Character-Aware Neural Language Models](https://arxiv.org/pdf/1508.06615.pdf) to split identifiers. 
+
+Train a basic Tensorflow neural net directly in the cluster, to learn how to split those identifiers
```
Comment (vmarkovtsev) [inline doc/neural_splitter_arch.md]: erase the whole cluster mention

File: doc/neural_splitter_arch.md [hunk 10]
```diff
@@ -0,0 +1,29 @@
+# Neural Token Splitter architecture description
+
+The goal of this feature is to train a neural network to learn how to split identifiers. Indeed, the rule-based `TokenParser` is not able to process some of them such as "`foobar`" or "`methodbase`". The main code is going to be stored in the algorithms directory: Class: `NeuralTokenSplitter`. The implementation plan is a follows.
+
+1. Use `utils.engine.create_engine` to initialize a spark session and the engine with `args.repositories` as input, siva format. We plan to process the dataset of 150k top starred repositories in any languages. 
+
+2. Code the new class named `CodeExtractor` in `transformers.basic` that is going to return both the source code as strings and their languages.
+
+3. Get the Token stream of identifiers thanks to [pygments.highlight](http://pygments.org/docs/quickstart/) with the `RawFormatter` and the language output by [enry](https://github.com/src-d/enry) as a lexer. Filter the token types with a callback. Design the code to easily switch between Babelfish and pygments.
```
Comment (vmarkovtsev) [inline doc/neural_splitter_arch.md]: ~thanks to~ using

**Comment by warenlg** [comment]
@vmarkovtsev  may I ask you to merge this one this one so that I could sign off next commit

**Comment by vmarkovtsev** [comment]
```
git commit --amend --no-edit --signoff
git push -f
```

**Comment by warenlg** [comment]
done

**Pull request merged**